src/OpenLoco/Windows/ScenarioSelect.cpp

```diff
@@ -0,0 +1,503 @@
+        self->widgets = _widgets;
+        self->enabled_widgets = (1 << widx::close) | (1 << widx::tab0) | (1 << widx::tab1) | (1 << widx::tab2) | (1 << widx::tab3) | (1 << widx::tab4);
+        self->initScrollWidgets();
+
+        self->colours[0] = Colour::black;
+        self->colours[1] = Colour::saturated_green;
+
+        self->var_846 = 0xFFFF;
+        self->var_85A = static_cast<int32_t>(0xFFFFFFFF);
+
+        initTabs(self);
+
+        // Select the last tab used, or the first available one.
+        uint8_t selectedTab = Config::get().scenario_selected_tab;
+        if (self->widgets[widx::tab0 + selectedTab].type == widget_type::none)
+        {
+            selectedTab = 0;
+            for (int i = 0; i < 5; i++)
+            {
+                if (self->widgets[widx::tab0 + i].type == widget_type::none)
+                {
+                    selectedTab = i;
+                    break;
+                }
+            }
+        }
+
+        self->current_tab = selectedTab;
+
+        initList(self);
+
+        return self;
+    }
+
+    // 0x00443995
+    static void prepareDraw(window* self)
+    {
+        self->activated_widgets &= ~((1 << widx::tab0) | (1 << widx::tab1) | (1 << widx::tab2) | (1 << widx::tab3) | (1 << widx::tab4));
+        self->activated_widgets |= (1ULL << (self->current_tab + static_cast<uint8_t>(widx::tab0)));
+    }
+
+    // 0x004439AF
+    static void draw(window* self, Gfx::drawpixelinfo_t* dpi)
+    {
+        Gfx::drawRectInset(dpi, self->x, self->y + 20, self->width, 41, self->colours[0], 0);
+
+        // Draw widgets.
+        self->draw(dpi);
+
+        static const string_id scenarioGroupIds[] = {
+            StringIds::scenario_group_beginner,
+            StringIds::scenario_group_easy,
+            StringIds::scenario_group_medium,
+            StringIds::scenario_group_challenging,
+            StringIds::scenario_group_expert,
+        };
+
+        // Draw tab captions.
+        for (int i = 0; i < 5; i++)
+        {
+            widget_t& widget = self->widgets[widx::tab0 + i];
+            if (widget.type == widget_type::none)
+                continue;
+
+            const auto offset = self->current_tab == i ? 1 : 0;
+            auto origin = Gfx::point_t(widget.mid_x() + self->x, widget.mid_y() + self->y - 3 - offset);
+            const string_id caption = scenarioGroupIds[i];
+
+            Gfx::drawStringCentredWrapped(dpi, &origin, 87, Colour::black, StringIds::wcolour2_stringid, &caption);
+        }
+
+        // Scenario selected?
+        if (self->info == 0 || self->info == 0xFFFFFFFF)
+            return;
+
+        using namespace scenariomgr;
+        auto scenarioInfo = reinterpret_cast<ScenarioIndexEntry*>(self->info);
+
+        // Check if current currency object needs to be changed.
+        registers regs;
+        regs.ebp = reinterpret_cast<int32_t>(&scenarioInfo->currency);
+        bool changeCurrency = (call(0x004720EB, regs) & (X86_FLAG_CARRY << 8)) == 0;
```

Comment:
This not longer crashes, but does not lead to reloading when expected (e.g. from dollar to pound). Fairly sure I have to check the carry flag, but…?

Reply:
`ObjectManager::getLoadedObjectIndex` expects an `object_index_entry` as its parameter, while we only have the header here. I suppose I could refactor that.